accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,502 @@
+        /// </summary>
+        protected abstract IEnumerable<Measurement<T>> Observe();
+    }
+}
+```
+
+### Measurement class
+
+```csharp
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A helper class used by the Observable instruments Observe method to repro the measurement values
+    /// with the associated tags.
+    /// </summary>
+    public struct Measurement<T> where T : unmanaged
+    {
+        /// <summary>
+        /// Construct the Measurement using the value and the list of tags.
+        /// We'll always copy the input list as this is not perf hot path.
+        /// </summary>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object>> tags) { throw null; }
+        public Measurement(T value, params KeyValuePair<string, object?>[] tags) { throw null; }
+
+        public ReadOnlySpan<KeyValuePair<string, object?>> Tags { get { throw null;  } }
+        public T Value { get; }
+    }
+}
+```
+
+### Non-observable Instruments
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// The counter is a non-observable Instrument that supports non-negative increments.
+    /// e.g. Number of completed requests.
+    /// </summary>
+    public class Counter<T> : Instrument<T> where T : unmanaged
+    {
+        public void Add(T measurement) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2,
+                            KeyValuePair<string, object> tag3) { throw null; }
+        public void Add(T measurement,
+                            params KeyValuePair<string, object>[] tags) { throw null; }
+    }
+
+    /// <summary>
+    /// UpDownCounter is a non-observable Instrument that supports increments and decrements.
+    /// e.g. Number of items in a queue.
+    /// </summary>
+    public class UpDownCounter<T> : Instrument<T> where T : unmanaged
+    {
+        public void Add(T measurement) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2) { throw null; }
+        public void Add(T measurement,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2,
+                            KeyValuePair<string, object> tag3) { throw null; }
+        public void Add(T measurement,
```

Comment:
This signature looks EXACTLY like Counter<T>. The main difference is that Add() takes a negative number in UpDownCounter<T>. For clarity, does it make sense in .NET to use Increment() and Decrement()?

PS: I know the OTel spec says Add().

Reply:
I prefer `Add` here as it will accept positive and negative values and that will be legitimate. But using Increment/Decrement and pass negative values that will be undesired. I know we do the same with Counter as it supposes to get positive values only but at least with UpDownCounter we are reducing this problem when we use `Add` :-)

Reply:
So, what do you do if I Add(-10) in a Counter<long>?

Reply:
Give it to @cijothomas and let him handle it 😄

Reply:
LOL. I agree. Chuck it as the SDK!